### PR TITLE
Flowchart Script

### LINKED ISSUE #43: Flowchart -> JSON Automation
Problem: There's a :chungus: amount of questions and answers in some of the flowcharts
Solution: Automate the process - given a lucidchart, spit out the relevant questions.json

Description:
LucidCharts can be made to export to a csv, that contains the relevant information necessary to convert into a JSON. We want to be able to download the csv of a LucidChart, plug it into a script, and have that script produce the appropriate JSON. This will probably need to be updated as more question types and fields are added, but for now it will just support the content present in #42.

Specs:
- [x] Script imports one or multiple csvs (If multiple, it should still consolidate into one object)
- [x] Script turns Process blocks into Questions
- [x] Script turns Arrow blocks into answers in the respective questions
- [x] The final object should be readable as type `Question[]`
  - [x] We can either store it into a JSON file, or **have the script run each time and store the csvs in the repo** which would eliminate the need for JSON files overall
- [x] Make compatible with #40 
  - [X] Continue question type
  - [x] Tooltip support

Expected Completion: 1-2 weeks
Resident Expert: #engineering-support

### PR DESCRIPTION
**Pull request by dillydally414**
Closes #43

### CODE REVIEW

**Review by dillydally414 — COMMENTED** [review]
You can run the test once using `npm run test`, or set into watch mode (it will re-run every time you update a file) using `npm run test:watch`. This test just checks that reading the animal form flowchart csv gives the appropriate `Question[]` output. Lmk if you have any questions.

File: components/DynamicForm/MyContinue.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import React, { useEffect } from 'react'
```
Comment (imanmoreira) [inline components/DynamicForm/MyContinue.tsx:1]: Notes for people reviewing: Ignore this, these are changes from the other PR

File: components/DynamicForm/ChooseFormType.tsx [hunk 2]
```diff
@@ -4,11 +4,13 @@ import { Question, Answer, AnswersKeys, answers } from '../../models'
 import { MyRadio } from './MyRadio'
 import { MyResult } from './MyResult'
 import { FormAnswer } from '../../utils/FormContext'
+import MyContinue from './MyContinue'
```
Comment (imanmoreira) [inline components/DynamicForm/ChooseFormType.tsx:7]: Notes for people reviewing: Ignore this file, these are changes from the other PR

File: components/DynamicForm/MyInput.tsx [hunk 3]
```diff
@@ -1,16 +1,18 @@
 import { FieldHookConfig, useField } from 'formik'
 import React, { ChangeEvent } from 'react'
```
Comment (imanmoreira) [inline components/DynamicForm/MyInput.tsx:2]: Notes for people reviewing: Ignore this, these are changes from the other PR

File: components/DynamicForm/MyRadio.tsx [hunk 4]
```diff
@@ -5,13 +5,16 @@ import { FormAnswer } from '../../utils/FormContext'
 import { RadioButton } from '../../components/FormStyles/RadioButton'
 import { QuestionText } from '../FormStyles/QuestionText'
 import QuestionLayout from '../FormStyles/QuestionLayout'
+import Tooltip from './Tooltip'
```
Comment (imanmoreira) [inline components/DynamicForm/MyRadio.tsx:8]: Notes for people reviewing: Ignore this, these are changes from the other PR

File: components/DynamicForm/MyResult.tsx [hunk 5]
```diff
@@ -5,35 +5,51 @@ import {
...
   label: string
   description?: string
+  questions: Question[]
 }
 
 // translates ID-based results to content-based results
-export function buildResults(formAnswers: {
-  [key: string]: FormAnswer
-}): FormAnswer[] {
+export function buildResults(
```
Comment (imanmoreira) [inline components/DynamicForm/MyResult.tsx:25]: Notes for people reviewing: Ignore this part, these are changes from the other PR

File: components/DynamicForm/Tooltip.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import React from 'react'
```
Comment (imanmoreira) [inline components/DynamicForm/Tooltip.tsx:1]: Notes for people reviewing: Ignore this, these are changes from the other PR

File: components/FormStyles/QuestionLayout.tsx [hunk 7]
```diff
@@ -3,6 +3,7 @@ import styled from 'styled-components'
 interface QuestionProps {
   questionText: JSX.Element
   input: JSX.Element
+  tooltip?: JSX.Element
```
Comment (imanmoreira) [inline components/FormStyles/QuestionLayout.tsx:6]: Notes for people reviewing: Ignore this file, these are changes from the other PR

File: constants/csv_parser.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  )
+  const idMap = new Map<number, number>()
+  filterOutTooltips.forEach((question: Question, index: number) => {
+    idMap.set(question.id, index)
+  })
+  const questions = filterOutTooltips.map((question: Question) => {
+    const newQuestionId = idMap.get(question.id)
+    return {
+      ...question,
+      id: newQuestionId === undefined ? -1 : newQuestionId,
```
Comment (imanmoreira) [inline constants/csv_parser.ts:98]: A neat little trick, instead of doing something like this you can do
newQuestionId || -1
or since you only defined a const because it's used twice in the above, 
idMap.get(question.id) || -1

File: constants/csv_parser.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+      answers: [
+        ...question.answers.map((answer: Answer) => {
+          const newRoute = idMap.get(answer.route)
```
Comment (imanmoreira) [inline constants/csv_parser.ts:101]: same thing for  newRoute as newQuestionId

**Review by imanmoreira — APPROVED** [review]
Most of my changes requested were nitpicks but take a look for code cleanliness

**Review by Nickdaco — COMMENTED** [review]
There's just a few areas that need more documentation because they are difficult to follow. but, `if it works it works`?

**Review by clue4 — COMMENTED** [review]
agreed with iman and nick, some more comments (especially on the parser) would be super helpful! otherwise i don't have anything else to add

File: pages/DynamicPOC/index.tsx [hunk 10]
```diff
@@ -63,22 +49,46 @@ const TitleText = styled.h1`
   font-size: large;
 `
 
-const DynamicPOC: React.FC = () => {
+const files = {
+  animalForm: '../../../constants/Animal Form.csv',
+  actualForm: '../../../constants/EdLaw Combined Flowchart.csv',
+}
+
+export const getStaticProps: GetStaticProps = (context) => {
```
[the remark this answers is not in the document]
Comment (dillydally414) [reply pages/DynamicPOC/index.tsx:57]: This is a funky Next.js way of passing props from server side into a component. Some of the modules used for reading the csv are backend only, which means if I try to call the csvToQuestionArray directly in the component it will throw an error. So, the workaround is using getStaticProps which will run this code server-side and then pass the result into DynamicPOC. Here's a link to the docs: https://nextjs.org/docs/basic-features/data-fetching/get-static-props

File: constants/csv_parser.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+type CsvType = {
+  Id: string
+  Name: string
+  'Line Source': string
+  'Line Destination': string
+  'Text Area 1': string
+  'Contained By': string
+}
+
```
[the remark this answers is not in the document]
Comment (dillydally414) [reply constants/csv_parser.ts:14]: Good point :skull:

File: constants/csv_parser.ts (hunk 8 above)
Comment (dillydally414) [reply constants/csv_parser.ts:98]: problem here is bc 0 is falsey so it converts 0 id to -1

File: pages/DynamicPOC/index.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (dillydally414) [reply pages/DynamicPOC/index.tsx:57]: I should clarify, this never gets called explicitly by us, but by exporting it Next knows to use it when building the pages

File: constants/csv_parser.ts [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+// The type containing columns that we care about in the Flowchart CSV
+type CsvType = {
+  Id: string
+  Name: string
+  'Line Source': string
+  'Line Destination': string
+  'Text Area 1': string
+  'Contained By': string
+}
```
Comment (VKong6019) [inline constants/csv_parser.ts:14]: hmm, my first thought was if these string fields could be converted into enums? i'd imagine this was prob considered before. if possible, i think having the fields like `Line Source` could be more readable since it seems like they're used often throughout here

i did quick internet investigation and found [this ](https://stackoverflow.com/questions/55142177/how-to-build-a-type-from-enum-values-in-typescript)
basically, would it be possible to convert the `type CsvType` to smth like this with [`Record`](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type)?
```suggestion
enum CsvKeys {
  ID: 'Id',
  NAME: 'Name',
  LINE_SOURCE: 'Line Source'
  ....
}

type CsvType = Record<CsvKeys, string>;
```

File: constants/csv_parser.ts [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+  ) {
+    return {
+      route: parseInt(answer['Line Destination']),
+    }
+  } else {
+    return {
+      content: answer['Text Area 1'],
+      route: parseInt(answer['Line Destination']),
+    }
+  }
```
Comment (VKong6019) [inline constants/csv_parser.ts:51]: this is a nit, so feel free to ignore if you'd like

seems like the if statement is only to add a single field, so i maybe do smth like

```suggestion
  const csvAnswer = { route: parseInt(answer['Line Destination']) };
  if (
    answer['Text Area 1'] !== 'CONTINUE' &&
    answer['Text Area 1'] !== 'TEXT'
  ) {
     csvAnswer['content'] = answer['Text Area 1'];
  }
  
  return csvAnswer;
```

Comment (VKong6019) [reply constants/csv_parser.ts:51]: with the power of de morgan's law :hellmo:

alternatively, you could also have the if statement say 
```suggestion
  if (['CONTINUE' , 'TEXT'].includes(answer['Text Area 1'])) {
  ...
  }
```
but up to you, whatever seems readable 👍

File: constants/csv_parser.ts [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+        | undefined = undefined
+      const tooltipText = tooltipTextMap.get(id)
+      const tooltipHoveredText = tooltipHoverTextMap.get(id)
+      if (tooltipText && tooltipHoveredText) {
+        tooltip = {
+          tooltipText: tooltipText,
+          tooltipHoveredText: tooltipHoveredText,
+        }
+      }
+      // eliminate funky newline characters in CSV
```
Comment (VKong6019) [inline constants/csv_parser.ts:170]: funky indeed

**Comment by VKong6019** [comment]
for the most part, looks pretty readable :D

File: constants/csv_parser.ts (hunk 12 above)
Comment (dillydally414) [reply constants/csv_parser.ts:14]: Wait this page also has so many other useful type things that I didn't know TypeScript had, TY VERA!
